### PR TITLE
[tests] fix flaky pattern in `test_generate_continue_from_past_key_values`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes the flakiness pattern behind a [common flaky test](https://app.circleci.com/insights/github/huggingface/transformers/workflows/run_tests/tests?branch=main), `tests/models/kosmos2/test_modeling_kosmos2.py::Kosmos2ModelTest::test_generate_continue_from_past_key_values`

flakiness pattern TL;DR: 
- on mixin tests, the models have random weights
- different input shapes can cause slightly different outputs (see [here](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535))
- in `generate`, tiny differences in the logits can lead to a different token, changing generation from that point onward
- equivalence checks for `generate` shouldn't naively look at the output sequences, but rather at the output sequences up to the point where they diverge, if they diverge 👉 corrects all equivalence checks using the right pattern (this pattern was already present in most tests, but a few ones still had the old pattern)

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1989,32 +1994,6 @@ def test_generate_continue_from_inputs_embeds(self):
                         )
                     )
 
-    @parameterized.expand([("offloaded",)])  # ("offloaded_static",) TODO: @raushan fixme in some models (eg T5)
-    @require_torch_accelerator
-    @pytest.mark.generate
-    def test_offloaded_cache_implementation(self, cache_implementation):
```
Comment (gante) [inline tests/generation/test_utils.py:1995]: this test doesn't make sense? tests that two `generate` calls with the same flags have the same output

We have tests for the offloaded cache in `tests/utils/test_cache_utils.py`

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1096,25 +1096,18 @@ def test_contrastive_generate_low_memory(self):
 
             # test output equality of low versus high memory
             model = model_class(config).to(torch_device).eval()
+            generate_kwargs = {
+                "top_k": 4,
+                "penalty_alpha": 0.6,
+                "max_new_tokens": self.max_new_tokens,
+                "use_cache": True,
+                "return_dict_in_generate": True,
+                "output_scores": True,
```
Comment (gante) [inline tests/generation/test_utils.py:1105]: (the correct equivalence check needs the scores)

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1862,22 +1855,29 @@ def test_generate_continue_from_past_key_values(self):
...
 
+            generate_kwargs = {
+                "pad_token_id": -1,
+                "eos_token_id": -1,
+                "forced_eos_token_id": None,
+                "encoder_no_repeat_ngram_size": 0,
+                "use_cache": True,
+                "do_sample": False,
+                "return_dict_in_generate": True,
+                "output_scores": True,
```
Comment (gante) [inline tests/generation/test_utils.py:1872]: (same, the correct equivalence check needs the scores)

**Review by zucchini-nlp — APPROVED** [review]
Nice!

File: tests/generation/test_utils.py (hunk 1 above)
Comment (gante) [reply tests/generation/test_utils.py:1995]: There are slow tests for all caches on `main`, [this PR](https://github.com/huggingface/transformers/pull/37394) adds a fast test for all caches 🤗

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -1929,6 +1932,8 @@ def test_generate_continue_from_inputs_embeds(self):
 
             if config.is_encoder_decoder:
                 self.skipTest(reason="This model is encoder-decoder")
+            # TODO (joao, raushan): the correct line below is `if not hasattr(config.get_text_config(), "use_cache")`,
+            # but it breaks a few models. Fix and then apply `_check_similar_generate_outputs` pattern
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1936]: Yup 👍 I've pushed the actual fix into the future, in case the test is not (much) flaky
